controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+		return
+	}
+
+	if m.snapshotCheckTaskQueue.NumRequeues(key) < snapshotHashMaxRetries {
+		logrus.Warnf("Error syncing snapshot check task %v: %v", key, err)
+		m.snapshotCheckTaskQueue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+
+	logrus.Warnf("Dropping hashing request of snapshot %v: %v", key, err)
+	m.snapshotCheckTaskQueue.Forget(key)
+}
+
+func (m *SnapshotMonitor) snapshotCheckWorker(id int) {
+	for m.processNextWorkItem(id) {
+	}
+}
+
+func (m *SnapshotMonitor) Close() {
+	m.logger.WithField("monitor", monitorName).Info("Closing snapshot monitor")
+
+	m.snapshotCheckTaskQueue.ShutDown()
+	m.checkScheduler.Stop()
+	m.quit()
+}
+
+func (m *SnapshotMonitor) Update(map[string]interface{}) error {
+	dataIntegrityCronJob, err := m.ds.GetSettingValueExisted(types.SettingNameSnapshotDataIntegrityCronJob)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameSnapshotDataIntegrityCronJob)
+	}
+
+	if dataIntegrityCronJob == m.existingDataIntegrityCronJob && m.checkScheduler.Len() > 0 {
+		return nil
+	}
+
+	if m.checkScheduler.Len() > 0 {
+		m.checkScheduler.Remove(m.checkSnapshots)
+	}
+
+	job, err := m.checkScheduler.Cron(dataIntegrityCronJob).Do(m.checkSnapshots)
+	if err != nil {
+		return errors.Wrap(err, "failed to schedule snapshot check job")
+	}
+
+	m.Lock()
+	m.existingDataIntegrityCronJob = dataIntegrityCronJob
+	m.Unlock()
+
+	m.checkScheduler.StartAsync()
+
+	m.logger.WithField("monitor", monitorName).Infof("Next snapshot check job will be executed at %v", job.NextRun())
+
+	return nil
+}
+
+func (m *SnapshotMonitor) GetData() (interface{}, error) {
+	m.RLock()
+	defer m.RUnlock()
+	return m.SnapshotMonitorStatus, nil
+}
+
+func (m *SnapshotMonitor) addToInProgressSnapshotCheckTasks(snapshotName string) bool {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	_, ok := m.inProgressSnapshotCheckTasks[snapshotName]
+	if ok {
+		m.logger.WithField("monitor", monitorName).Debugf("snapshot %s is being checked", snapshotName)
+		return false
+	}
+	m.inProgressSnapshotCheckTasks[snapshotName] = struct{}{}
+
+	return true
+}
+
+func (m *SnapshotMonitor) deleteFromInProgressSnapshotCheckTasks(snapshotName string) {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	delete(m.inProgressSnapshotCheckTasks, snapshotName)
+}
+
+func (m *SnapshotMonitor) RunTask(arg interface{}) error {
+	task, ok := arg.(snapshotCheckTask)
+	if !ok {
+		return fmt.Errorf("failed to assert value: %v", arg)
+	}
+
+	if !m.addToInProgressSnapshotCheckTasks(task.snapshotName) {
+		return nil
+	}
+	defer m.deleteFromInProgressSnapshotCheckTasks(task.snapshotName)
+
+	engine, err := m.ds.GetVolumeCurrentEngine(task.volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get engine for volume %v", task.volumeName)
+	}
+
+	if err := m.checkVolumeNotInMigration(engine.Spec.VolumeName); err != nil {
```

Comment:
May need to check other operations like rebuilding and purge

Reply:
Addressed. Thank you.